KAN-77 feat: 가짜 리뷰 검사 기능

nuyeo opened the pull request:
### 🔗 관련 이슈
KAN-77

### 📌 작업 내용
이 PR에서 추가된 기능 또는 수정 사항에 대해 간략히 설명해주세요.

- 프론트엔드에서 검사할 URL을 보내면 AI 서버로 검사 요청을 보냄
- AI서버에서 결과를 받아와 redis에 2시간 동안 캐싱
- redis에 캐싱된 URL에 대한 요청을 받으면 캐싱된 결과로 응답
- AI서버에서 아직 결과를 받지 못했을 경우 임시 로딩 상태를 응답
- AI서버로 검사 요청을 보낼 때 임의의 고유 requestId 생성

### 🧑‍💻 작업 유형
- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] 코드 리팩토링
- [ ] 문서 수정
- [ ] CI/CD 변경

### 📷 관련 이미지
- 사용자가 리뷰 검사 요청
<img width="1278" alt="스크린샷 2024-11-09 오후 6 25 54" src="https://github.com/user-attachments/assets/57ada5a6-05de-40d6-92b2-931fd606f229">

- AI서버에서 보낸 리뷰 검사 결과 캐싱
<img width="1278" alt="스크린샷 2024-11-09 오후 6 27 34" src="https://github.com/user-attachments/assets/21f88946-fe3d-4701-8c20-58bc4376f8fa">

- 캐싱된 리뷰 검사 결과 조회
<img width="1278" alt="스크린샷 2024-11-09 오후 6 27 58" src="https://github.com/user-attachments/assets/ce59f4e9-132f-4a55-9134-d5b19947ea19">

### 🚨 주의 사항
- FE에서 검사 요청 보낸 후 폴링 방식으로 AI 서버에서 결과 받아왔는지 확인하는 로직 추가 필요
- 로컬에서 실행할 경우 redis 설정 필요

### ✅ 체크 리스트
- [x] 코드 리뷰어가 이해할 수 있도록 주석 및 설명을 추가했습니다.
- [x] 테스트를 추가하거나 업데이트 했습니다.
- [x] 이 변경 사항이 다른 기능에 영향을 미치는지 확인했습니다.

nuyeo committed:
KAN-77 feat: BaseEntity 추가

모든 Entity에 공통으로 들어가는 칼럼을 BaseEntity로 정의

nuyeo committed:
KAN-77 모든 엔티티에 BaseEntity 상속 추가

nuyeo committed:
KAN-77 Spring Data JPA에서 Auditing 활성화

BaseEntity의 @CreatedDate와 @LastModifiedDate 사용을 위해 Auditing 기능 활성화

nuyeo committed:
KAN-77 chore: Redis 의존성 추가

nuyeo committed:
KAN-77 chore: Swagger 의존성 추가

nuyeo committed:
KAN-77 feat: swagger-ui 추가

nuyeo committed:
KAN-77 chore: 폴더 구조 변경

nuyeo committed:
KAN-77 feat: Redis 설정 추가

nuyeo committed:
KAN-77 feat: 블로그 URL을 받아 요청하는 DTO 정의

nuyeo committed:
KAN-77 feat: 가짜 리뷰 검사 비즈니스 로직 구현

nuyeo committed:
KAN-77 feat: 가짜 리뷰 검사 시 Redis 캐싱 로직 추가

nuyeo committed:
KAN-77 feat: 임의 AI서버 URL 정의

nuyeo committed:
KAN-77 chore: 테스트 코드 작성을 위한 의존성 추가

nuyeo committed:
KAN-77 test: 가짜 리뷰 검사 테스트 코드 추가

nuyeo committed:
KAN-77 test: 캐시에 값이 없을 때 테스트 기대값이 notNull인지 확인하도록 변경

sall6550 reviewed (approved):
고생하셨습니다~

nuyeo committed:
KAN-77 feat: AI서버에서 응답을 받아오는 엔드포인트 추가

nuyeo committed:
KAN-77 feat: AI 서버에 리뷰 검사 요청 시 요청ID 생성 기능 추가

nuyeo committed:
KAN-77 feat: 검사 결과를 조회하는 엔드포인트 추가

konempty commented on build.gradle:
```diff
@@ -41,6 +45,12 @@ dependencies {
+    implementation 'org.springdoc:springdoc-openapi-starter-webmvc-ui:2.6.0'
+
+    // redis
+    implementation 'org.springframework.boot:spring-boot-starter-data-redis'
```
엇 ㅋㅋㅋ 26번째 줄에 이미 레디스가 선언되어 있는것 같아요!

<img width="652" alt="스크린샷 2024-11-09 오후 11 08 33" src="https://github.com/user-attachments/assets/2d9c5b70-b8c0-45fb-8815-fe34e46b65d2">

nuyeo replied:
헉 ㅋㅋㅋㅋ 중복이 있었네요..

konempty commented on src/main/java/com/spring/be/config/RedisConfig.java:
```diff
@@ -0,0 +1,25 @@
+public class RedisConfig {
+
+    @Bean
+    public RedisTemplate<String, String> redisTemplate(RedisConnectionFactory connectionFactory) {
```
Spring data redis를 사용하는 경우 별다른 요청이 없다면 기본적으로 Lettuce라는 라이브러리를 기반으로 동작하는데
복잡한 서비스, 혹은 분산락이 필요한 서비스에서는 
Reddison이라는 라이브러리를 쓰는게 훨씬 까다롭긴 하지만 성능 측면에서는 좋다는건 알고 계시면 좋을듯합니다~!
그렇다고 이번 프로젝트에서 Redisson을 쓰라는건 아니고 그런 선택지가 있다는 것만 알아두시면 좋을것 같아요!

konempty commented on src/main/java/com/spring/be/entity/BaseEntity.java:
```diff
@@ -0,0 +1,28 @@
+    @Column(name = "updated_at")
+    private LocalDateTime updatedAt;
+
+    @Column(name = "deleted_at")
+    private LocalDateTime deletedAt;
```
오호 deleteAt이 존재 한다는건 Soft delete를 사용하실 계획이신가요??

nuyeo replied:
피드백이나 후기 작성에 유저 정보가 포함되어 있어서 softDelete를 사용하려고 했습니다!

konempty commented on src/main/java/com/spring/be/entity/BaseEntity.java:
```diff
@@ -0,0 +1,28 @@
+public abstract class BaseEntity {
+
+    @CreatedDate
+    @Column(name = "created_at", updatable = false)
```
다른 엔티티에서는 컬럼명을 명시적으로 지정해주진 않는것 같은데
요 베이스 엔티티에는 지정해주는 이유가 있을까요??

nuyeo replied:
다른 엔티티는 @sall6550님이 구현하셨는데  BaseEntity만 제가 구현해서 차이가 있었습니다 😅
통일성을 위해 베이스 엔티티에서도 명시적 지정 제거했습니다!

konempty commented on src/main/java/com/spring/be/reviewcheck/controller/ReviewCheckController.java:
```diff
@@ -0,0 +1,44 @@
+    public ResponseEntity<String> receiveAIResponse(@RequestBody ReviewCheckResult result) {
+        // AI 서버의 응답을 캐시 및 데이터베이스에 저장
+        reviewCheckService.cachedReviewCheckResult(result.getRequestId(), result);
+        return ResponseEntity.ok("AI response received and processed.");
```
응답이 성공했고 뭔가 내릴정보는 딱히 없지만 뭔가는 내려줘야 할것 같아서 String을 내리시는거라면
Ok(200)대신 NoContent(204)를 내리시는게 좋을것 같아요~!

nuyeo replied:
오 정확히 캐치하셨군요..! 204로 내리도록 수정하겠습니다~!

konempty commented on src/main/java/com/spring/be/reviewcheck/dto/ReviewCheckRequest.java:
```diff
@@ -0,0 +1,11 @@
+import lombok.Setter;
+
+@Getter @Setter
+public class ReviewCheckRequest {
+    @NotBlank
+    private String blogUrl;
+}
```
kotlin에는 data class라는 이름으로 getter와 setter등을 자동으로 구현해주는 기능이 있는데
java에는 해당 개념에 대응하는 기능으로 record class라는게 있어요!
그걸 Request나 Response, 혹은 불변 DTO등에 사용하면 롬복에 대한 의존도를 줄일 수있는데
이번기회에 공부해두시면 좋을것 같아요!

nuyeo replied:
감사합니다 관련해서 공부하도록 하겠습니다!

konempty commented on src/main/java/com/spring/be/reviewcheck/service/ReviewQueueService.java:
```diff
@@ -0,0 +1,117 @@
+    public ObjectMapper objectMapper;
+
+    @Autowired
+    public ReviewQueueService(RedisTemplate<String, String> redisTemplate, RedisCacheUtil redisCacheUtil,
+                              ReviewCheckResultRepository reviewCheckResultRepository, RestTemplate restTemplate,
+                              ObjectMapper objectMapper) {
+        this.redisTemplate = redisTemplate;
+        this.redisCacheUtil = redisCacheUtil;
+        this.reviewCheckResultRepository = reviewCheckResultRepository;
+        this.restTemplate = restTemplate;
+        this.objectMapper = objectMapper;
+    }
+
```
생성자로 주입받을 거라면 
각 프로퍼티에서 @ Autowired 어노테이션은 없애도 괜찮지 않을까요??

nuyeo replied:
생성자 주입 방식으로 변경해두었습니다!

konempty commented on src/main/resources/application.properties:
```diff
@@ -14,4 +14,14 @@ jwt.expirationMs=3600000
+springdoc.api-docs.enabled=true
+
+# AI server
+ai.server.url=http://localhost:6000
```
<img width="382" alt="스크린샷 2024-11-09 오후 11 00 55" src="https://github.com/user-attachments/assets/dfdbb35c-9d41-4aa1-9270-2dab69762973">

파일 끝 개행은 넣어두는게 좋을것 같아요~!

nuyeo replied:
개행 빠진 부분도 다시 체크해서 추가하겠습니다!

konempty commented on src/test/java/com/spring/be/reviewcheck/service/ReviewCheckServiceTest.java:
```diff
@@ -0,0 +1,121 @@
+import org.mockito.MockitoAnnotations;
+//import org.springframework.boot.test.context.SpringBootTest;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
```
와일드 카드 임포트는 예상치 못한 이슈를 발생시킬 수 있으므로 사용은 지양하시는게 좋습니다!

nuyeo replied:
명시적 임포트로 변경했습니다!

konempty commented:
디코로 질문 주신 주기적인 Polling에 대해 사용사례가 하나 떠올랐는데
카카오 로그인을 할때 QR로그인을 선택하면
웹 페이지는 사용자가 QR로그인을 완료했는지 확인하기위해 1초에 한번씩 서버에 로그인 여부를 물어보고(check_token_state.json)
로그인이 완료되었다면 로그인 처리를 하는식으로 구현되어있어요!

Polling을 쓰는게 기능과 구현에 따라 나쁜것만은 아닐수도 있다는걸 알고 계시면 좋을 듯합니다~!

<img width="1654" alt="스크린샷 2024-11-09 오후 10 23 29" src="https://github.com/user-attachments/assets/69022cb0-011b-4e6d-a90a-d8b577e1c770">

konempty commented on src/main/java/com/spring/be/config/RedisConfig.java:
```diff
@@ -0,0 +1,25 @@
+        RedisTemplate<String, String> template = new RedisTemplate<>();
+        template.setConnectionFactory(connectionFactory);
+
+        template.setKeySerializer(new StringRedisSerializer());
+        template.setValueSerializer(new StringRedisSerializer());
+        template.setHashKeySerializer(new StringRedisSerializer());
+        template.setHashValueSerializer(new StringRedisSerializer());
```
요렇게 전부 StringRedisSerializer쓸 계획이라면 RedisTemplate대신에
StringRedisTemplate를 쓰는게 좋을것 같아요~! ㅋㅋㅋ

nuyeo replied:
object에서 string으로 변경하면서 RedisTemplate를 그대로 사용하고 있었네요.. 수정했습니다!

dayaelee commented:
전반적으로 잘 작성하신것 같습니다! 고생하셨습니다!

dayaelee reviewed (approved):
전반적으로 잘 작성하신것 같습니다! 고생하셨습니다!

nuyeo committed:
fix: 중복 선언된 dependency 제거

nuyeo committed:
refactor: 의존성 주입을 필드 주입에서 생성자 주입으로 변경

- @Autowired 어노테이션을 통한 필드 주입 방식을 제거하고 생성자 주입 방식으로 변경
- 불변성 보장 및 final 키워드 추가로 초기화 강제

nuyeo committed:
chore: 파일 하단 개행 및 주석 추가

nuyeo committed:
refactor: 와일드 카드 import에서 명시적 개별 import로 변경

nuyeo commented:
@konempty 카카오 로그인에서 polling을 사용하는지는 전혀 몰랐네요! 좋은 예시 감사합니다 :)

nuyeo committed:
test: 테스트 코드에 requestId 추가

nuyeo committed:
refactor: RedisTemplate를 StringRedisTemplate으로 변경

- string 직렬화를 위한 StringRedisSerializer 설정 제거

nuyeo committed:
refactor: AI서버 응답 시 HTTP 204 No Content 상태 코드로 응답하도록 수정

- 응답 데이터가 필요 없다면 200 OK 대신 204 No Content 사용

konempty commented on src/main/java/com/spring/be/reviewcheck/controller/ReviewCheckController.java:
```diff
@@ -0,0 +1,44 @@
+    public ResponseEntity<String> receiveAIResponse(@RequestBody ReviewCheckResult result) {
+        // AI 서버의 응답을 캐시 및 데이터베이스에 저장
+        reviewCheckService.cachedReviewCheckResult(result.getRequestId(), result);
+        return ResponseEntity.noContent().build();
```
@nuyeo 
요 경우 리턴값으로 주는것 보다는
메소드에
@ ResponseStatus(NO_CONTENT) 어노테이션을 붙히시는게 간편합니다~!

nuyeo replied:
수정했습니다!

nuyeo committed:
refactor: BaseEntity에서 명시적 컬럼명 지정 제거

nuyeo committed:
refactor: AI 서버 응답 상태 코드를 리턴값에서 @ResponseStatus 어노테이션으로 변경
